crypto/src/hash.rs

```diff
@@ -84,15 +83,53 @@ impl Hash {
 
 impl From<Hash> for [u8; Hash::LENGTH] {
     #[inline]
-    fn from(Hash(bytes): Hash) -> Self {
-        bytes
+    fn from(hash: Hash) -> Self {
+        #[allow(unsafe_code)]
+        // SAFETY: `Hash` and `[u8; Hash::LENGTH]` have the same memory layout
+        unsafe {
+            core::mem::transmute(hash)
+        }
     }
 }
 
 impl AsRef<[u8; Hash::LENGTH]> for Hash {
     #[inline]
     fn as_ref(&self) -> &[u8; Hash::LENGTH] {
-        &self.0
+        #[allow(unsafe_code, trivial_casts)]
+        // SAFETY: `Hash` and `[u8; Hash::LENGTH]` have the same memory layout
+        unsafe {
+            &*((self as *const Self).cast::<[u8; Self::LENGTH]>())
+        }
+    }
+}
+
+impl Serialize for Hash {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let hash: &[u8; Self::LENGTH] = self.as_ref();
+        hash.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Hash {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
```

Comment:
This is incredibly dangerous. What if we set the LSB to 0 by accident, and the remaining hash is non-zero.

Reply:
We would silently set LSB to 1 inside `prehashed` function, however i agree that we should explicitly return error if hash is not valid.